app/channel.go

```diff
@@ -1899,7 +1899,15 @@ func (a *App) GetChannelUnread(channelID, userID string) (*model.ChannelUnread,
 	if channelUnread.NotifyProps[model.MARK_UNREAD_NOTIFY_PROP] == model.CHANNEL_MARK_UNREAD_MENTION {
 		channelUnread.MsgCount = 0
 	}
-
+	// if we don't have MentionCountRoot yet, calculate it now and retry
+	if channelUnread.MentionCountRoot == nil {
+		err := a.populateMentionCountRoot(channelUnread.TeamId, userID)
+		if err != nil {
+			return nil, model.NewAppError("GetChannelUnread", "app.channel.get_unread.app_error", nil, err.Error(), http.StatusInternalServerError)
+		}
+		// retry
+		return a.GetChannelUnread(channelID, userID)
```

Comment:
This recursive self-call behavior seems potentially dangerous, because it could lead to infinite recursion in case `populateMentionCountRoot` never assigns a non-nil value to MentionCountRoot. I looked at `populateMentionCountRoot` and it seems there's only a few places where a value is assigned, which are nested in some ifs and fors.

Are you absolutely sure that there's no edge case in which `channelUneread.MentionCountRoot` remains `nil` even after an successful populateMentionCountRoot call?

Reply:
@zefhemel I'll rewrite it as non-recursive